File: CSharp/SpaceVIL/SpaceVIL/Items/TreeView.cs
using System;
using System.Drawing;
using System.Linq;
using System.Collections.Generic;
using SpaceVIL.Core;
using SpaceVIL.Common;
using SpaceVIL.Decorations;

namespace SpaceVIL
{
    /// <summary>
    /// TreeView is special container designed to show content as tree view structure.
    /// <para/> Extended from SpaceVIL.ListBox.
    /// <para/> Supports all events except drag and drop.
    /// </summary>
    public class TreeView : ListBox
    {
        /// <summary>
        /// Event that is invoked when tree view need to sort content.
        /// </summary>
        public EventCommonMethod EventSortTree;

        /// <summary>
        /// Disposing TreeView resources if it was removed.
        /// <para/> Notice: This method is mainly for overriding only. SpaceVIL calls 
        /// this method if necessary and no need to call it manually.
        /// </summary>
        public override void Release()
        {
            EventSortTree = null;
        }

        internal int MaxWrapperWidth = 0;

        internal TreeItem _root; //nesting level = 0

        /// <summary>
        /// Setting the root (head) SpaceVIl.TreeItem is visible or invisible.
        /// </summary>
        /// <param name="value">True: if you want root SpaceVIl.TreeItem to be visible.
        /// False: if you want root SpaceVIl.TreeItem to be invisible.</param>
        public void SetRootVisible(bool value)
        {
            if (_root == null)
                return;
            _root.SetVisible(value);
            GetWrapper(_root).SetVisible(value);
            //reset all paddings for content
            List<IBaseItem> list = GetListContent();
            if (list != null)
            {
                foreach (var item in list)
                {
                    TreeItem tmp = item as TreeItem;
                    if (tmp != null)
                        tmp.ResetIndents();
                }
            }
            UpdateElements();
        }

        /// <summary>
        /// Returns True if root (head) SpaceVIl.TreeItem is visible otherwise returns False.
        /// </summary>
        /// <returns>True: if root SpaceVIl.TreeItem is visible.
        /// False: if root SpaceVIl.TreeItem is invisible.</returns>
        public bool IsRootVisible()
        {
            if (_root == null)
                return false;
            return _root.IsVisible();
        }

        /// <summary>
        /// Setting text to root (head) SpaceVIl.TreeItem of TreeView.
        /// </summary>
        /// <param name="text">Text for root.</param>
        public void SetRootText(String text)
        {
            if (_root == null)
                return;
            _root.SetText(text);
        }

        /// <summary>
        /// Getting text of root (head) SpaceVIl.TreeItem of TreeView.
        /// </summary>
        /// <returns>Text of root.</returns>
        public String GetRootText()
        {
            if (_root == null)
                return "";
            return _root.GetText();
        }

        /// <summary>
        /// Getting root (head) SpaceVIl.TreeItem of TreeView.
        /// </summary>
        /// <returns>Root as SpaceVIl.TreeItem.</returns>
        public TreeItem GetRootItem()
        {
            return _root;
        }

        /// <summary>
        /// Setting new root (head) SpaceVIl.TreeItem for TreeView.
        /// </summary>
        /// <param name="rootTreeItem">New root as SpaceVIl.TreeItem.</param>
        public void SetRootItem(TreeItem rootTreeItem)
        {
            if (_root != null)
                RemoveItem(_root);
            AddItem(rootTreeItem);
        }

        private static int count = 0;

        /// <summary>
        /// Default TreeView constructor.
        /// </summary>
        public TreeView()
        {
            SetItemName("TreeView_" + count);
            count++;
            _root = new TreeItem(TreeItemType.Branch, "root");
            _root.SetItemName("Root");

            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.TreeView)));
            EventSortTree += OnSortTree;

            SetHScrollBarPolicy(VisibilityPolicy.AsNeeded);
        }

        /// <summary>
        /// Initializing all elements in the TreeView. 
        /// <para/> Notice: This method is mainly for overriding only. SpaceVIL calls 
        /// this method if necessary and no need to call it manually.
        /// </summary>
        public override void InitElements()
        {
            base.InitElements();
            _root._treeViewContainer = this;
            _root.SetRoot(true);
            _root.GetIndicator().SetToggled(true);
            base.AddItem(_root);
            SetRootVisible(false);
            // _root.ResetIndents();
            MaxWrapperWidth = GetWrapper(_root).GetMinWidth();
        }

        internal void RefreshWrapperWidth()
        {
            foreach (SelectionItem wrp in GetArea()._mapContent.Values)
            {
                wrp.SetMinWidth(MaxWrapperWidth);
                wrp.GetContent().SetMinWidth(MaxWrapperWidth);
            }
        }

        internal void RefreshTree(TreeItem prev, TreeItem item)
        {
            List<IBaseItem> list = GetListContent();
            int index = GetListContent().IndexOf(prev) + 1;
            int nestLev = item._nestingLevel;
            while (index < list.Count)
            {
                if (((TreeItem)list[index])._nestingLevel <= nestLev)
                    break;
                index++;
            }
            InsertItem(item, index);
            item.ResetIndents();
            // item.OnToggleHide(true);
            UpdateElements();
        }

        /// <summary>
        /// Adding all elements in the list area of TreeView from the given list.
        /// </summary>
        /// <param name="content">List of items as IEnumerable&lt;IBaseItem&gt;</param>
        public override void SetListContent(IEnumerable<IBaseItem> content)
        {
            GetArea().Clear();
            foreach (IBaseItem item in content)
            {
                AddItem(item);
            }
        }

        /// <summary>
        /// Updating all TreeView inner items.
        /// </summary>
        public override void UpdateElements()
        {
            MaxWrapperWidth = 0;
            foreach (SelectionItem wrp in GetArea()._mapContent.Values)
            {
                if (!wrp.IsVisible())
                    continue;
                TreeItem item = wrp.GetContent() as TreeItem;
                if (item != null)
                {
                    item.ResetIndents();
                    int actualWrpWidth = GetWidth() - (GetArea().GetPadding().Left + GetArea().GetPadding().Right)
                            - (wrp.GetMargin().Left + wrp.GetMargin().Right);
                    wrp.SetWidth(actualWrpWidth - (VScrollBar.IsDrawable() ? VScrollBar.GetWidth() : 0));
                }
            }
            base.UpdateElements();
        }

        private void OnSortTree()
        {
            if (_root == null)
                return;
            //sorting
            List<TreeItem> outList = new List<TreeItem>();
            outList.Add(_root);
            outList.AddRange(SortHelper(_root));

            SetListContent(outList.Select(_ => _ as IBaseItem).ToList());
        }

        internal SelectionItem GetWrapper(TreeItem item)
        {
            return GetArea()._mapContent[item];
        }

        private List<TreeItem> SortHelper(TreeItem item)
        {
            List<TreeItem> tmpList = item.GetChildren();
            tmpList.Sort(CompareInAlphabet);
            List<TreeItem> outList = new List<TreeItem>();
            foreach (TreeItem ti in tmpList)
            {
                outList.Add(ti);
                if (ti.GetItemType() == TreeItemType.Branch)
                    outList.AddRange(SortHelper(ti));
            }

            return outList;
        }

        internal int CompareInAlphabet(TreeItem ti1, TreeItem ti2)
        {
            if (ti1.GetItemType() != ti2.GetItemType())
            {
                if (ti1.GetItemType() == TreeItemType.Branch) return -1;
                else return 1;
            }

            return ti1.GetText().ToLower().CompareTo(ti2.GetText().ToLower());
        }

        /// <summary>
        /// Adding a node to the TreeView.
        /// </summary>
        /// <param name="item">Node as SpaceVIL.Core.IBaseItem.</param>
        public override void AddItem(IBaseItem item)
        {
            if (_root == null)
            {
                TreeItem tmp = item as TreeItem;
                if (tmp != null)
                {
                    _root = tmp;
                    _root._treeViewContainer = this;
                    _root.SetRoot(true);
                    _root.GetIndicator().SetToggled(true);
                    base.AddItem(_root);
                    SetRootVisible(false);

                    MaxWrapperWidth = GetWrapper(_root).GetMinWidth();
                }
                //exception: ///////
            }
            else
                _root.AddItem(item);
        }

        /// <summary>
        /// Setting style of the TreeView. 
        /// <para/> Inner styles: "area", "vscrollbar", "hscrollbar", "menu".
        /// </summary>
        /// <param name="style">Style as SpaceVIL.Decorations.Style.</param>
        public override void SetStyle(Style style)
        {
            if (style == null)
                return;
            base.SetStyle(style);
            //additional
        }

        /// <summary>
        /// Removing all items from the list area of TreeView.
        /// </summary>
        public override void Clear()
        {
            _root.RemoveChildren();
            base.Clear();
            base.AddItem(_root);
            SetRootVisible(IsRootVisible());
            MaxWrapperWidth = GetWrapper(_root).GetMinWidth();

        }

        /// <summary>
        /// Removing the specified item from the list area of TreeView.
        /// </summary>
        /// <param name="item">Item as SpaceVIL.Core.IBaseItem.</param>
        /// <returns>True: if the removal was successful. 
        /// False: if the removal was unsuccessful.</returns>
        public override bool RemoveItem(IBaseItem item)
        {
            if (item == null)
                return false;
            if (item.Equals(_root))
            {
                _root.RemoveChildren();
                _root = null;
            }
            return base.RemoveItem(item);
        }

        /// <summary>
        /// Sorting TreeView nodes in internal list area starting with root (head).
        /// </summary>
        public void SortTree()
        {
            SortBrunch(_root);
        }

        /// <summary>
        /// Sorting part of TreeView content starting with specified branch node.
        /// </summary>
        /// <param name="branch">Branch node as SpaceVIL.TreeItem.</param>
        public void SortBrunch(TreeItem branch)
        {
            if (branch == null)
                return;
            if (branch.GetItemType().Equals(TreeItemType.Leaf))
            {
                return; // Либо сделать, чтобы сортировалась родительская ветвь?
            }

            List<IBaseItem> list = GetArea().GetItems();
            Dictionary<int, List<SelectionItem>> savedMap = new Dictionary<int, List<SelectionItem>>();

            int indFirst = list.IndexOf(GetWrapper(branch)) + 1;
            int nestLev = branch._nestingLevel + 1;
            int indLast = indFirst;
            int maxLev = nestLev;

            while (indLast < list.Count)
            {
                SelectionItem si = ((SelectionItem)list[indLast]);
                int stiLev = ((TreeItem)si.GetContent())._nestingLevel;
                if (stiLev < nestLev)
                    break;

                if (maxLev < stiLev)
                    maxLev = stiLev;

                if (!savedMap.ContainsKey(stiLev))
                {
                    List<SelectionItem> l1 = new List<SelectionItem>();
                    savedMap.Add(stiLev, l1);
                }

                savedMap[stiLev].Add(si);

                list.RemoveAt(indLast);
            }

            for (int i = nestLev; i <= maxLev; i++)
            {
                List<SelectionItem> siList = savedMap[i];
                if (siList == null)
                    continue;

                foreach (SelectionItem selIt in siList)
                {
                    TreeItem curItm = ((TreeItem)selIt.GetContent());
                    TreeItem parItm = curItm.GetParentBranch();
                    int parNum = list.IndexOf(GetWrapper(parItm));

                    int ind = parNum;

                    for (int ii = parNum + 1; ii < list.Count; ii++)
                    {
                        TreeItem tmpItm = (TreeItem)((SelectionItem)list[ii]).GetContent();
                        if (tmpItm._nestingLevel <= parItm._nestingLevel)
                            break;

                        int compRes = CompareInAlphabet(tmpItm, curItm);
                        if (compRes > 0)
                            break;
                        ind = ii;
                    }

                    list.Insert(ind + 1, selIt);
                }
            }

            GetArea().SetContent(list);

            UpdateElements();
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/TreeView.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Common.DefaultsService;
import com.spvessel.spacevil.Core.EventCommonMethod;
import com.spvessel.spacevil.Core.IBaseItem;
import com.spvessel.spacevil.Decorations.Style;
import com.spvessel.spacevil.Flags.VisibilityPolicy;
import com.spvessel.spacevil.Flags.TreeItemType;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;

/**
 * TreeView is special container designed to show content as tree view
 * structure.
 * <p>
 * Extended from com.spvessel.spacevil.ListBox.
 * <p>
 * Supports all events except drag and drop.
 */
public class TreeView extends ListBox {
    public EventCommonMethod eventSortTree = new EventCommonMethod();

    /**
     * Disposing TreeView resources if it was removed.
     * <p>
     * Notice: This method is mainly for overriding only. SpaceVIL calls this method
     * if necessary and no need to call it manually.
     */
    @Override
    public void release() {
        eventSortTree.clear();
    }

    TreeItem _root; // nesting level = 0

    int maxWrapperWidth = 0;

    /**
     * Setting the root (head) com.spvessel.spacevil.TreeItem is visible or
     * invisible.
     * 
     * @param visible True: if you want root com.spvessel.spacevil.TreeItem to be
     *                visible. False: if you want root
     *                com.spvessel.spacevil.TreeItem to be invisible.
     */
    public void setRootVisible(boolean visible) {
        if (_root == null)
            return;
        _root.setVisible(visible);
        getWrapper(_root).setVisible(visible);
        // reset all paddings for content
        List<IBaseItem> list = getListContent();
        if (list != null) {
            for (IBaseItem item : list) {
                if (item instanceof TreeItem) {
                    if (item.isVisible()) {
                        TreeItem tmp = (TreeItem) item;
                        tmp.resetIndents();
                    }
                }
            }
        }
        updateElements();
    }

    /**
     * Returns True if root (head) com.spvessel.spacevil.TreeItem is visible
     * otherwise returns False.
     * 
     * @return True: if root com.spvessel.spacevil.TreeItem is visible. False: if
     *         root com.spvessel.spacevil.TreeItem is invisible.
     */
    public boolean isRootVisible() {
        if (_root == null)
            return false;
        return _root.isVisible();
    }

    /**
     * Setting text to root (head) com.spvessel.spacevil.TreeItem of TreeView.
     * 
     * @param text Text for root.
     */
    public void setRootText(String text) {
        if (_root == null)
            return;
        _root.setText(text);
    }

    /**
     * Getting text of root (head) com.spvessel.spacevil.TreeItem of TreeView.
     * 
     * @return Text of root.
     */
    public String getRootText() {
        if (_root == null)
            return "";
        return _root.getText();
    }

    /**
     * Getting root (head) com.spvessel.spacevil.TreeItem of TreeView.
     * 
     * @return Root as com.spvessel.spacevil.TreeItem.
     */
    public TreeItem getRootItem() {
        return _root;
    }

    /**
     * Setting new root (head) com.spvessel.spacevil.TreeItem for TreeView.
     * 
     * @param rootTreeItem New root as com.spvessel.spacevil.TreeItem.
     */
    public void setRootItem(TreeItem rootTreeItem) {
        if (_root != null)
            removeItem(_root);
        addItem(rootTreeItem);
    }

    private static int count = 0;

    /**
     * Default TreeView constructor.
     */
    public TreeView() {
        setItemName("TreeView_" + count);
        count++;
        _root = new TreeItem(TreeItemType.Branch, "root");
        _root.setItemName("root");

        setStyle(DefaultsService.getDefaultStyle(TreeView.class));
        eventSortTree.add(this::onSortTree);

        setHScrollBarPolicy(VisibilityPolicy.AsNeeded);
    }

    /**
     * Initializing all elements in the TreeView.
     * <p>
     * Notice: This method is mainly for overriding only. SpaceVIL calls this method
     * if necessary and no need to call it manually.
     */
    @Override
    public void initElements() {
        super.initElements();
        _root._treeViewContainer = this;
        _root.setRoot(true);
        _root.getIndicator().setToggled(true);
        super.addItem(_root);
        setRootVisible(false);

        // _root.resetIndents();
        maxWrapperWidth = getWrapper(_root).getMinWidth();
    }

    void refreshWrapperWidth() {
        for (SelectionItem wrp : getArea()._mapContent.values()) {
            wrp.setMinWidth(maxWrapperWidth);
            wrp.getContent().setMinWidth(maxWrapperWidth);
        }
    }

    void refreshTree(TreeItem prev, TreeItem item) {
        List<IBaseItem> list = getListContent();
        int index = getListContent().indexOf(prev) + 1;
        int nestLev = item._nestingLevel;
        while (index < list.size()) {
            if (((TreeItem) list.get(index))._nestingLevel <= nestLev)
                break;
            index++;
        }
        insertItem(item, index);
        item.resetIndents();
        // item.onToggleHide(true);
        updateElements();
    }

    /**
     * Adding all elements in the list area of TreeView from the given list.
     * 
     * @param content List of items as
     *                List&lt;com.spvessel.spacevil.Core.IBaseItem&gt;
     */
    @Override
    public void setListContent(List<IBaseItem> content) {
        getArea().clear();

        for (IBaseItem item : content) {
            addItem(item);
        }
    }

    /**
     * Updating all TreeView inner items.
     */
    @Override
    public void updateElements() {
        maxWrapperWidth = 0;
        for (SelectionItem wrp : getArea()._mapContent.values()) {
            if (!wrp.isVisible())
                continue;

            if (wrp.getContent() instanceof TreeItem) {
                ((TreeItem) wrp.getContent()).resetIndents();
                int actualWrpWidth = getWidth() - (getArea().getPadding().left + getArea().getPadding().right)
                        - (wrp.getMargin().left + wrp.getMargin().right);
                wrp.setWidth(actualWrpWidth - (vScrollBar.isDrawable() ? vScrollBar.getWidth() : 0));
            }
        }
        super.updateElements();
    }

    private void onSortTree() {
        if (_root == null)
            return;
        // sorting
        List<TreeItem> outList = new LinkedList<>();
        outList.add(_root);
        outList.addAll(sortHelper(_root));

        List<IBaseItem> list = new LinkedList<>();
        for (TreeItem var : outList) {
            list.add(var);
        }
        setListContent(list);
    }

    SelectionItem getWrapper(TreeItem item) {
        return getArea()._mapContent.get(item);
    }

    private List<TreeItem> sortHelper(TreeItem item) {
        List<TreeItem> tmpList = item.getChildren();

        Collections.sort(tmpList, new CompareInAlphabet());

        List<TreeItem> outList = new LinkedList<>();
        for (TreeItem ti : tmpList) {
            outList.add(ti);
            if (ti.getItemType() == TreeItemType.Branch)
                outList.addAll(sortHelper(ti));
        }

        return outList;
    }

    class CompareInAlphabet implements Comparator<TreeItem> {
        @Override
        public int compare(TreeItem ti1, TreeItem ti2) {
            if (ti1.getItemType() != ti2.getItemType()) {
                if (ti1.getItemType() == TreeItemType.Branch)
                    return -1;
                else
                    return 1;
            }
            return ti1.getText().compareToIgnoreCase(ti2.getText());
        }
    }

    Comparator<TreeItem> getComparator() {
        return new CompareInAlphabet();
    }

    /**
     * Adding a node to the TreeView.
     * 
     * @param item Item as com.spvessel.spacevil.Core.IBaseItem.
     */
    @Override
    public void addItem(IBaseItem item) {
        if (_root == null) {
            if (item instanceof TreeItem) {
                _root = (TreeItem) item;
                _root._treeViewContainer = this;
                _root.setRoot(true);
                _root.getIndicator().setToggled(true);
                super.addItem(_root);
                setRootVisible(false);

                maxWrapperWidth = getWrapper(_root).getMinWidth();
            }
            // exception: ///////
        } else
            _root.addItem(item);
    }

    /**
     * Setting style of the TreeView.
     * <p>
     * Inner styles: "area", "vscrollbar", "hscrollbar", "menu".
     * 
     * @param style Style as com.spvessel.spacevil.Decorations.Style.
     */
    @Override
    public void setStyle(Style style) {
        if (style == null)
            return;
        super.setStyle(style);
        // additional

    }

    /**
     * Removing all items from the list area of TreeView.
     */
    @Override
    public void clear() {
        _root.removeChildren();
        super.clear();
        super.addItem(_root);
        setRootVisible(isRootVisible());
        maxWrapperWidth = getWrapper(_root).getMinWidth();
    }

    /**
     * Removing the specified item from the list area of TreeView.
     * 
     * @param item Item as com.spvessel.spacevil.Core.IBaseItem.
     * @return True: if the removal was successful. False: if the removal was
     *         unsuccessful.
     */
    @Override
    public boolean removeItem(IBaseItem item) {
        if (item == null)
            return false;
        if (item.equals(_root)) {
            _root.removeChildren();
            _root = null;
        }
        return super.removeItem(item);
    }

    /**
     * Sorting TreeView nodes in internal list area starting with root (head).
     */
    public void sortTree() {
        sortBrunch(_root);
    }

    /**
     * Sorting part of TreeView content starting with specified branch node.
     * 
     * @param branch Branch node as com.spvessel.spacevil.TreeItem.
     */
    public void sortBrunch(TreeItem branch) {
        if (branch == null)
            return;
        if (branch.getItemType().equals(TreeItemType.Leaf)) {
            return; // Либо сделать, чтобы сортировалась родительская ветвь?
        }

        List<IBaseItem> list = getArea().getItems();
        Map<Integer, List<SelectionItem>> savedMap = new HashMap<>();

        int indFirst = list.indexOf(getWrapper(branch)) + 1;
        int nestLev = branch._nestingLevel + 1;
        int indLast = indFirst;
        int maxLev = nestLev;

        while (indLast < list.size()) {
            SelectionItem si = ((SelectionItem) list.get(indLast));
            int stiLev = ((TreeItem) si.getContent())._nestingLevel;
            if (stiLev < nestLev)
                break;

            if (maxLev < stiLev)
                maxLev = stiLev;

            if (!savedMap.containsKey(stiLev)) {
                List<SelectionItem> l1 = new LinkedList<>();
                savedMap.put(stiLev, l1);
            }

            savedMap.get(stiLev).add(si);

            list.remove(indLast);
        }

        Comparator<TreeItem> comp = getComparator();

        for (int i = nestLev; i <= maxLev; i++) {
            List<SelectionItem> siList = savedMap.get(i);
            if (siList == null)
                continue;

            for (SelectionItem selIt : siList) {
                TreeItem curItm = ((TreeItem) selIt.getContent());
                TreeItem parItm = curItm.getParentBranch();
                int parNum = list.indexOf(getWrapper(parItm));

                int ind = parNum;

                for (int ii = parNum + 1; ii < list.size(); ii++) {
                    TreeItem tmpItm = (TreeItem) ((SelectionItem) list.get(ii)).getContent();
                    if (tmpItm._nestingLevel <= parItm._nestingLevel)
                        break;

                    int out = comp.compare(tmpItm, curItm);
                    if (out > 0)
                        break;
                    ind = ii;
                }

                list.add(ind + 1, selIt);
            }
        }

        getArea().setContent(list);

        updateElements();
    }
}